File: src/person_service/scripts/person_client.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

# 该例程将请求/show_person服务，服务数据类型person_service::s_c_person

import sys
import rospy
from person_service.srv import s_c_person, s_c_personRequest

def person_client():

	# ROS节点初始化
    rospy.init_node('person_client')

	# 发现/show_person服务后，创建一个服务客户端，连接名为/show_person的service
    rospy.wait_for_service('/show_person')
    try:
        # 创建一个客户端
        person_client = rospy.ServiceProxy('/show_person', s_c_person)

		# 请求服务调用，输入请求数据
        response = person_client("Tom", 20, s_c_personRequest.male)

        return response.result #将结果返回给main函数并显示

    except rospy.ServiceException, e:
        print "Service call failed: %s"%e

if __name__ == "__main__":

	#服务调用并显示调用结果
    print "Show person result : %s" %(person_client())

File: src/person_service/src/person_client.cpp
#include "ros/ros.h"
#include "person_service/s_c_person.h"//头文件名字与数据类型名字一样

int main(int argc,char**argv)
{
    //初始化ros节点
    ros::init(argc,argv,"person_client");//节点名字"person_client"

    //创建节点句柄
    ros::NodeHandle node;

    //等待服务器启动，创建一个服务客户端,类型为person_service，名字为s_c_person（.h的文件名），连接名为/show_person的service
    ros::service::waitForService("/show_person");
    ros::ServiceClient person_client=node.serviceClient<person_service::s_c_person>("/show_person");

    //初始化person_service::s_c_person的请求数据
    person_service::s_c_person srv;
    srv.request.name="lzx";
    srv.request.age=20;
    srv.request.sex=person_service::s_c_person::Request::male;

    //请求服务调用
    ROS_INFO("Call service to show person[name:%s, age:%d,sex:%d]",
    srv.request.name.c_str(),srv.request.age,srv.request.sex);

    person_client.call(srv);//请求客户端

    //显示服务结果
    ROS_INFO("show person result:%s",srv.response.result.c_str());

    return 0;
}